transcendence-app/src/user/user.controller.ts

```diff
@@ -1,39 +1,63 @@
 import {
-  BadRequestException,
+  Body,
   Controller,
   Get,
   NotFoundException,
   Param,
+  ParseBoolPipe,
+  ParseIntPipe,
+  ParseUUIDPipe,
   Post,
   Query,
+  Request as GetRequest,
+  UseGuards,
 } from '@nestjs/common';
 import { UserService } from './user.service';
 import { UserEntity } from './user.entity';
 import { UserDto } from './dto/user.dto';
-import { plainToClass } from 'class-transformer';
-import { validate } from 'class-validator';
+import { AuthenticatedGuard } from '../shared/guards/authenticated.guard';
+import { Request } from 'express';
+import { ApiForbiddenResponse, ApiTags } from '@nestjs/swagger';
 
+const MAX_USER_ENTRIES_PER_PAGE = 20;
+@ApiTags('user')
+@ApiForbiddenResponse()
+@UseGuards(AuthenticatedGuard)
 @Controller('user')
 export class UserController {
   constructor(private userService: UserService) {}
 
-  @Post('new?')
-  async addUser(@Query() query: UserDto): Promise<UserEntity> {
-    const validatedUser = plainToClass(UserDto, query);
-    const errors = await validate(validatedUser, {
-      skipMissingProperties: false,
-      forbidUnknownValues: true,
-    });
-    if (errors.length > 0) {
-      throw new BadRequestException();
-    }
-    return this.userService.findOneOrCreate(query);
+  @Post()
+  async addUser(@Body() user: UserDto): Promise<UserEntity> {
+    return this.userService.findOneOrCreate(user);
   }
 
-  @Get(':id')
-  async getUserById(@Param('id') param: string): Promise<UserEntity> {
-    const result = this.userService.retrieveUserWithId(Number(param));
-    if (result === undefined) throw new NotFoundException();
+  @Get('me')
+  getCurrentUser(@GetRequest() req: Request) {
+    return req.user;
+  }
+
+  @Get('users')
+  getUsers() {
+    return this.userService.retrieveUsers();
+  }
+
+  @Get('usernames?')
+  getUsernames(
+    @Query('take', ParseIntPipe) take: number,
+    @Query('skip', ParseIntPipe) skip: number,
+    @Query('sort', ParseBoolPipe) sort: boolean,
+  ) {
+    take = take > MAX_USER_ENTRIES_PER_PAGE ? MAX_USER_ENTRIES_PER_PAGE : take;
```

Comment:
suggestion: We could remove this max limit and let the callers of the API decide the limit depending on their needs.

Reply:
I don't know if there is a max retrieval we could configure in DB but I'd say is a good point to limit queries sizes since this directly affects performance of our system

Reply:
So, should we keep this max limit or remove it?

Reply:
what do you think @gbudau? I haven't researched how to solve this but I think is not a good idea to let the client request as much data as it wants :cry: Maybe there's a cleaner way we can solve this

Reply:
Great point! It's a valid and a good reason to limit the query size for performance reasons :+1: